src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java

```diff
@@ -0,0 +1,320 @@
+                worklist.enqueue(out.node);
+            }
+        }
+
+        private void visitBinary(Node node, Node left, Node right, binding_irnode.ir_opcode opcode) {
+            Map<Integer, List<Node>> nodeMap;
+            List<Node> rhsList;
+            if (!binaryNodeCache.containsKey(left.getNr())) {
+                // add new mapping for this entry
+                nodeMap = new HashMap<>();
+                binaryNodeCache.put(left.getNr(), nodeMap);
+                rhsList = new ArrayList<>();
+                nodeMap.put(right.getNr(), rhsList);
+            } else {
+                nodeMap = binaryNodeCache.get(left.getNr());
+                if (nodeMap.containsKey(right.getNr())) {
+                    System.out.println("already had value for " + left + ", " + right);
+                    rhsList = nodeMap.get(right.getNr());
+                    // if they have the same type just return the original node
+                    for (var n : rhsList) {
+                        if (n.getOpCode() == opcode) {
+                            // can exchange node with already found in list
+                            nodeValues.put(node, n);
+                            // enqueue outs
+                            for (var out : BackEdges.getOuts(node)) {
+                                worklist.enqueue(out.node);
+                            }
+                            return;
+                        }
+                    }
+                } else {
+                    rhsList = new ArrayList<>();
+                    nodeMap.put(right.getNr(), rhsList);
+                }
+            }
+
+            // add our own node to be found in later iterations
+            rhsList.add(node);
+        }
+
+        @Override
+        public void visit(Add node) {
+            visitBinary(node, node.getLeft(), node.getRight(), binding_irnode.ir_opcode.iro_Add);
+        }
+
+        @Override
+        public void visit(And node) {
+            visitBinary(node, node.getLeft(), node.getRight(), binding_irnode.ir_opcode.iro_And);
+        }
+
+        @Override
+        public void visit(Cmp node) {
+            // do nothing
+        }
+
+        @Override
+        public void visit(Cond node) {
+            // do nothing
+        }
+
+        @Override
+        public void visit(Const node) {
+            if (node.getMode().equals(Mode.getIs())) {
+                int val = node.getTarval().asInt();
+                if (constCache.containsKey(val)) {
+                    Node newConst = constCache.get(val);
+                    nodeValues.put(node, newConst);
+                } else {
+                    constCache.put(val, node);
+                }
+            }
+        }
+
+        @Override
+        public void visit(Conv node) {
+            visitUnary(node, node.getOp(), binding_irnode.ir_opcode.iro_Conv);
+        }
+
+        @Override
+        public void visit(Div node) {
```

Comment:
Do you consider the memory input of `Div` and `Mod` in any way? The language report does state that behavior of division by zero is undefined, but we can't just ignore the resulting exceptions either (at least so far we haven't).